src/modules/mc_pos_control/MulticopterPositionControl.cpp

```diff
@@ -281,179 +233,66 @@ void MulticopterPositionControl::Run()
-			// Example:
-			// If the desired setpoint is position-setpoint, _local_pos_sp will contain
-			// position-, velocity- and thrust-setpoint where the velocity- and thrust-setpoint were generated by the PositionControlller.
-			// If the desired setpoint has a velocity-setpoint only, then _local_pos_sp will contain valid velocity- and thrust-setpoint, but the position-setpoint
-			// will remain NAN. Given that the PositionController cannot generate a position-setpoint, this type of setpoint is always equal to the input to the
-			// PositionController.
+			// Publish internal position control setpoints
+			// on top of the input/feed-forward setpoints these containt the PID corrections
+			// This message is used by other modules (such as Landdetector) to determine vehicle intention.
 			vehicle_local_position_setpoint_s local_pos_sp{};
 			local_pos_sp.timestamp = time_stamp_now;
 			_control.getLocalPositionSetpoint(local_pos_sp);
-
-			// Publish local position setpoint
-			// This message will be used by other modules (such as Landdetector) to determine vehicle intention.
 			_local_pos_sp_pub.publish(local_pos_sp);
 
-			// Inform FlightTask about the input and output of the velocity controller
-			// This is used to properly initialize the velocity setpoint when onpening the position loop (position unlock)
-			_flight_tasks.updateVelocityControllerIO(Vector3f(local_pos_sp.vx, local_pos_sp.vy, local_pos_sp.vz),
-					Vector3f(local_pos_sp.acceleration));
-
-			vehicle_attitude_setpoint_s attitude_setpoint{};
-			attitude_setpoint.timestamp = time_stamp_now;
-			_control.getAttitudeSetpoint(attitude_setpoint);
-
-			// publish attitude setpoint
+			// Publish attitude setpoint output
 			// It's important to publish also when disarmed otheriwse the attitude setpoint stays uninitialized.
 			// Not publishing when not running a flight task
 			// in stabilized mode attitude setpoints get ignored
 			// in offboard with attitude setpoints they come from MAVLink directly
+			vehicle_attitude_setpoint_s attitude_setpoint{};
+			attitude_setpoint.timestamp = time_stamp_now;
+			_control.getAttitudeSetpoint(attitude_setpoint);
 			_vehicle_attitude_setpoint_pub.publish(attitude_setpoint);
 
-			_wv_dcm_z_sp_prev = Quatf(attitude_setpoint.q_d).dcm_z();
-
-			// if there's any change in landing gear setpoint publish it
-			landing_gear_s landing_gear = _flight_tasks.getGear();
-
-			if (landing_gear.landing_gear != _old_landing_gear_position
-			    && landing_gear.landing_gear != landing_gear_s::GEAR_KEEP) {
-
-				landing_gear.timestamp = hrt_absolute_time();
-				_landing_gear_pub.publish(landing_gear);
-			}
-
-			_old_landing_gear_position = landing_gear.landing_gear;
-
 		} else {
 			// reset the numerical derivatives to not generate d term spikes when coming from non-position controlled operation
 			_vel_x_deriv.reset();
```

Comment:
This looks wrong (resetting if `trajectory_setpoint` didn't update). Given the module is scheduled on `vehicle_local_position` publications we should probably be resetting it if the module itself hasn't run for a while and you might as well check `vehicle_local_position.v_xy_valid` as well.

I'd still like to consider filtering and publishing the control data externally so that we always have exactly what was being used.

Reply:
Yes, the intention was to not change too many things which gets non-rebasable/-reviewable. Originally this wasn't wrong since the position controller ran when a flight task ran and therefore the derivative had to start again when the position controller ran the next time. With this pr the assumption stays but it's certainly not the best way anymore to probe for the case and also the possible cases may change in the future.

Reply:
It should be fine for now if the current expectation is that a flight task will publish a `trajectory_setpoint` for every single consumed `vehicle_local_position` and both `flight_task_manager` and `mc_pos_control` and running at the same limited rate.